drivers/at_cmd/at_cmd.c

```diff
@@ -0,0 +1,455 @@
+
+	return client;
+}
+
+struct at_cmd_client *at_cmd_client_init(at_cmd_handler_t handler)
+{
+	struct at_cmd_client *client = NULL;
+
+	__ASSERT_NO_MSG(handler);
+
+	k_mutex_lock(&clients_mutex, K_FOREVER);
+
+	for (size_t i = 0; i < ARRAY_SIZE(clients); i++) {
+		if (clients[i]) {
+			continue;
+		}
+
+		client = client_add(i, handler);
+		break;
+	}
+
+	k_mutex_unlock(&clients_mutex);
+
+	if (!client) {
+		LOG_ERR("Could not register AT command client.");
+	}
+
+	return client;
+}
+
+static void client_free(size_t client_idx)
+{
+	k_free(&clients[client_idx]);
+	clients[client_idx] = NULL;
+}
+
+void at_cmd_client_uninit(struct at_cmd_client *client)
+{
+	size_t client_idx;
+
+	__ASSERT_NO_MSG(client);
+
+	k_mutex_lock(&clients_mutex, K_FOREVER);
+
+	client_idx = client_idx_from_id(client->id);
+	if (client_idx == INVALID_CLIENT_IDX) {
+		LOG_ERR("Client could not be removed, client ID not found.");
+	} else {
+		client_free(client_idx);
+	}
+
+	k_mutex_unlock(&clients_mutex);
+}
+
+int at_cmd_send(struct at_cmd_client *client, struct at_cmd_msg *cmd)
+{
+	struct at_cmd_fifo_item *fifo_cmd_data;
+	char *buf;
+
+	if (cmd->len > CMD_MAX_LEN) {
+		LOG_ERR("AT command length exceeds maximum allowed length.");
+		return -EMSGSIZE;
+	}
+
+	/* Allocate memory for the command struct. */
+	fifo_cmd_data = k_malloc(sizeof(struct at_cmd_fifo_item));
+	if (fifo_cmd_data == NULL) {
+		LOG_ERR("Memory could not be allocated for the command.");
+		return -ENOMEM;
+	}
+
+	/* Allocate memory for the comamnd buffer. */
+	buf = k_malloc(cmd->len);
+	if (buf == NULL) {
+		k_free(fifo_cmd_data);
+		LOG_ERR("Memory could not be allocated for command buffer.");
+		return -ENOMEM;
+	}
+
+	fifo_cmd_data->cmd.client = client;
+	fifo_cmd_data->cmd.timestamp = k_uptime_get_32();
+	fifo_cmd_data->cmd.buf = buf;
+	fifo_cmd_data->cmd.len = cmd->len;
+
+	memcpy(fifo_cmd_data->cmd.buf, cmd->buf, cmd->len);
+	memcpy(client->cmd, cmd->buf, cmd->len);
+
+	k_fifo_put(&at_cmd_fifo, fifo_cmd_data);
+
+	return 0;
+}
+
+static bool cmd_is_done(char *buf, size_t len)
+{
+	return strstr(buf, AT_CMD_OK) || strstr(buf, AT_CMD_ERROR);
+}
+
+/* @brief Function to get index of client with pending command.
+ * @return Client index on success or 0xBEEF on error.
+ */
+static size_t client_cmd_pending_get_idx(void)
+{
+	size_t i;
+
+	for (i = 0; i < ARRAY_SIZE(clients); i++) {
+		if (!clients[i]) {
+			continue;
+		}
+
+		if (clients[i]->cmd_pending) {
+			return i;
+		}
+	}
+
+	return INVALID_CLIENT_IDX;
+}
+
+static bool is_cesq_notification(char *buf, size_t len)
+{
+	/* Performing quick screening to see if it's worth doing more thorough
+	 * and cycle-demanding parsing of the received sring to determine if
+	 * it's a supported notification type.
+	 */
+	if ((len < AT_CMD_LEN(AT_CMD_CESQ_RESP)) ||
+	    (buf[4] != AT_CMD_CESQ_RESP[4])) {
+		return false;
+	}
+
+	return strstr(buf, AT_CMD_CESQ_RESP) ? true : false;
+}
+
+static bool is_cesq_subscribe(char *buf, size_t len)
```

Comment:
I am not sure if the way notification subscriptions is currently implemented will scale very well. The are for instance notifications coming from the modem that doesn't really need to firstly issue some specific commands, so it's not like a command-response pair. I think it would be better to separate command issuing and notification subscription into separate parts. There are situations where I want to subscribe on notifications without issuing a specific command, there may be multiple clients that want to subscribe on the same notification (without issuing a command first), ther may be clients that want to cancel a subscription.

Reply:
Agree, this subscription shows hos it _can_ be done, but don't think it's a good solution the more I look at it. I was considering using the event manager for subscriptions and notifications. Do you think that would be an ok solution? The event manager allows for one client to be prioritized, so in the event of for example an SMS, one client's callback could always be called first.

Reply:
If I understand what you are trying to do correctly I don't think the AT command interface is a good place to put it. I think you should abstract it away and provide a notification/subscription service on top of an shared socket. So that you can tell the driver that you want to subscribe on signal strength notification, phone service updates, or received SMSes, etc 

For instance to receive SMSes you must send this commands:
`AT+CEREG=2
 AT+CFUN=1
 AT+CNMI=3,2,0,1
 `
but these commands may also be used for other purposes so you can say that a client that issues any of this commands want to receive SMSes. If you have an interface on to of the AT command interface you can create a function to let the user receive SMSes, and that function can check that the modem is in the right state. This will also more easily let the driver manage conflicting states. The third argument that this will decouple the clients from the the AT command interface as will make the move to some other device that may have a different interface than AT command.

Reply:
Yes, that's a much better way to do it

Reply:
![screenshot_20190215_090018](https://user-images.githubusercontent.com/12154119/52842669-1eeb2a00-3100-11e9-9f1c-16d5c64713b7.png)